(GH-1781) Add ResourceInstance data type (#1786)
puppetlabs/bolt: documentation/experimental_features.md

  any way. It is only used to store information about a resource.

### Creating `ResourceInstance` objects

#### `Target.set_resources()`

The recommended way to create `ResourceInstance` objects is by setting them directly on
a `Target` object using the `Target.set_resources` function. This function can be used to 
set one or more resources on a target at a time. You can read more about this function in
the [Bolt plan functions reference documentation](plan_functions.md#set_resources).

The `Target.set_resources` function can set existing `ResourceInstance` objects on a target,
or take hashes of parameters to create new `ResourceInstance` objects and automatically set
them on a target.

When setting resources using a hash of parameters, you can pass either a single hash or an
array of hashes:

> **Note:** When passing a data hash to `Target.set_resources`, the `target` parameter
  is **optional**. If the `target` parameter is not specified, the function automatically
  sets the target to the target the function is called on.

> **Note:** If the `target` parameter is any target other than the one you are setting the

>   resource on, Bolt will raise an error.

Reviewer: Let's move these notes below the codeblock

```ruby
$init_hash = {
  'target'        => Optional[Target],
  'type'          => Variant[String[1], Type[Resource]],
  'title'         => String[1],
  'state'         => Optional[Hash[String[1], Data]],
  'desired_state' => Optional[Hash[String[1], Data]],
  'events'        => Optional[Array[Hash[String[1], Data]]]
}

$target.set_resources(
  Variant[Hash[String[1], Any], Array[Hash[String[1], Any]]] init_hash
)
```

> When setting resources using existing `ResourceInstance`s, you can pass either a single

Reviewer: ```suggestion
When setting resources using existing `ResourceInstance`s, you can pass either a single
```
This is more of a style thing, but I think the plural `ResourceInstance`s is going to look weird once it's rendered on the site. Maybe rephrase to "existing `ResourceInstance` objects"

> `ResourceInstance` or an array of `ResourceInstance`s.

Reviewer: ```suggestion
`ResourceInstance` or an array of `ResourceInstance` objects.
```

> **Note:** If the target for a `ResourceInstance` does not match the target it is being set
  on, Bolt will raise an error.

```ruby
$resource = ResourceInstance.new(...)

$target.set_resources(
  Variant[ResourceInstance, Array[ResourceInstance]] resource
)
```

> A target can only have a single instance of a given resource. If a duplicate resource is set
> on a target, the `state` and `desired_state` of the duplicate resource will be shallow merged
> with that of the existing resource, while any `events` for the duplicate resource will be
> added to the events for the existing resource.

Reviewer: ```suggestion
A target can only have a single instance of a given resource. If you set a duplicate resource
on a target, Bolt shallow merges the `state` and `desired_state` of the duplicate resource
with the `state` and `desired state` of the existing resource and adds any `events` for the 
duplicate resource to the existing resource.
```

| --- | --- | --- |
| `type` | The target that the resource is for. | `Target` |
| `type` | The [type of the resource](https://puppet.com/docs/puppet/latest/type.html). This can be either the stringified name of the resource type or the actual type itself. For example, both `"file"` and `File` are acceptable. | `Variant[String[1], Type[Resource]]` |
| `title` | The title, or [namevar](https://puppet.com/docs/puppet/latest/type.html#namevars-and-titles), of the resource. | `String[1]` |
| `state` | The _observed state_ of the resource. This is the point-in-time state of the resource when it is queried. | `Hash[String[1], Data]` |
| `desired_state` | The _desired state_ of the resource. This is the state that you want the resource to be in. | `Hash[String[1], Data]` |
| `events` | Resource events that are generated from reports. | `Array[Hash[String[1], Data]]` | 

A `ResourceInstance` is identified by its `target`, `type`, and `title`. As such, these
three parameters _must_ be specified when creating a new `ResourceInstance` and are
immutable. Since `Target.set_resources` will automatically set the `target` when passing
a hash of parameters, the `target` parameter can be ommitted.

> The `state`, `desired_state`, and `events` parameters are optional when creating a
> `ResourceInstance`. If `state` and `desired_state` are not specified they will default to
> empty hashes, while `events` will default to an empty array. Each of these attributes can
> be modified during the lifetime of the `ResourceInstance` using [the data type's
> functions](bolt_types_reference.md#resourceinstance).

Reviewer: ```suggestion
The `state`, `desired_state`, and `events` parameters are optional when creating a
`ResourceInstance`. If you do not specify `state` and `desired_state`, they default to
empty hashes, while `events` defaults to an empty array. You can modify each of 
these attributes during the lifetime of the `ResourceInstance` using [the data type's
functions](bolt_types_reference.md#resourceinstance).
```

### Functions

> The `ResourceInstance` data type has several built-in functions. These range from accessing
> the object's attributes to modifying and overwriting state. A full list of the available
> functions can be found in the [data type reference 
> documentation](bolt_types_reference.md#resourceinstance).

Reviewer: ```suggestion
The `ResourceInstance` data type has several built-in functions. These range from accessing
the object's attributes to modifying and overwriting state. For a full list of the available
functions, see [Bolt data types](bolt_types_reference.md#resourceinstance).
```
